pkg/applicationserver/applicationserver.go

```diff
@@ -290,7 +290,7 @@ func (as *ApplicationServer) processUp(ctx context.Context, up *ttnpb.Applicatio
 		return nil
 	}
 
-	if err := as.publishUp(ctx, up); err != nil {
+	if err := as.publishUp(as.FromRequestContext(ctx), up); err != nil {
```

Comment:
For the record - the context here is from the RPC call that NS makes - it will not outlive the publishing process.

Reply:
Then why would you decouple this?

In fact, if the NS wants to cancel the publishing process, it should do so via the context.

Reply:
This PR targets the AS link removal branch - the NS uses unary RPCs to push uplinks here. The context here is automatically cancelled as soon as the RPC callback is finished.

Reply:
Yes, I know, but we use context decoupling in places where it gets passed on to another goroutine, i.e. when sending it as part of `ContextualApplicationUp` on a channel. As long as that doesn't happen, there's no need to decouple the context, and in fact we shouldn't.

Reply:
But that's exactly what `publishUp` does indirectly:

`publishUp` calls `Publish` on the local distributor:
https://github.com/TheThingsNetwork/lorawan-stack/blob/e1431e984880da5c89da4f6ae3b109e320ec415e/pkg/applicationserver/applicationserver.go#L303-L304

`localDistributor` is a local distributor:
https://github.com/TheThingsNetwork/lorawan-stack/blob/e1431e984880da5c89da4f6ae3b109e320ec415e/pkg/applicationserver/applicationserver.go#L123

`localDistributor` `Publish` calls `subscriptionSet.Publish` and `subscriptionMap.Publish`:
https://github.com/TheThingsNetwork/lorawan-stack/blob/e1431e984880da5c89da4f6ae3b109e320ec415e/pkg/applicationserver/distribution/local_distributor.go#L42-L53

`subscriptionMap` is just a `map[ttnpb.ApplicationIdentifiers]*subscriptionSet`, and `subscriptionSet.Publish`  will create the `ContextualApplicationUp`:
https://github.com/TheThingsNetwork/lorawan-stack/blob/e1431e984880da5c89da4f6ae3b109e320ec415e/pkg/applicationserver/distribution/subscription_set.go#L98-L111

Fundamentally each time you do `publishUp` the uplink will end up in another goroutine. This is the whole reason for which I proposed that the decoupling is done as part of `publishUp` in the first place.

Reply:
Exactly for the code above, you should decouple the context as late as possible.

The caller wants to be able to cancel the publish, like canceling the call context that is used here:

https://github.com/TheThingsNetwork/lorawan-stack/blob/e1431e984880da5c89da4f6ae3b109e320ec415e/pkg/applicationserver/distribution/subscription_set.go#L104

If you do what you did, i.e. derive from the AS context in `publishUp()`, then this `ctx` is the component context already. How would you cancel waiting for having the message sent on a channel that way?

Therefore, call `FromRequestContext()` as late as possible, i.e. only when setting `Context` of `ContextualApplicationUp`, and use the request context for everything else.

Reply:
Ok, I understand what you mean now.